scipy/stats/stats.py

```diff
@@ -7469,6 +7470,254 @@ def brunnermunzel(x, y, alternative="two-sided", distribution="t",
     return BrunnerMunzelResult(wbfn, p)
 
 
+def _confint_lowerbound(n, quantile, confidence):
+    r"""
+    Compute the lower bound for a one-sided confidence interval
+    for a given
+    - quantile (0<`quantile`<1)
+    - confidence level (0<`confidence`<1)
+    - number of samples `n`.
+
+    Returns the largest index of the sample being a valid lower bound,
+    or `None` if there are not enough samples to derive one.
+
+    Used by the public function confint_quantile().
+
+    .. versionadded:: 1.6.0
+    """
+
+    # compute all probabilities from the binomial distribution for
+    # the quantile of interest
+    bd = binom(n, quantile)
+
+    # the lower bound is the last index before the invert survival
+    # function value for the target confidence level
+    lb = bd.isf(confidence) - 1
+
+    if lb < 0:  # isf returns -1 if there are no matching index
+        return None
+    else:
+        return int(lb)
+
+
+def confint_quantile(x, quantile, confidence, type='one-sided'):
+    r"""Compute non-parametric confidence intervals for any quantile.
+
+    This function implements a non-parametric approach to compute
+    confidence intervals for quantiles. The approach is attributed
+    to Thompson [1]_ and later proven to be applicable to any set
+    of i.i.d. samples [2]_. The computation is based on the observation
+    that the probability of a quantile :math:`q` to be larger than any
+    sample :math:`x_m (1\leq m \leq N)` can be computed as
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = 1 - \sum_{k=0}^{m-1} \binom{N}{k}
+        q^k(1-q)^{N-k}
+
+    Furthermore, these probabilities are symmetric, which allows to compute
+    both upper and lower bounds from the same computation:
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = \mathbb{P}(x_{N-m+1} \geq 1-q).
+
+    The function computes confidence intervals for a given quantile and
+    confidence level, based on `x` which is either a set of samples
+    (one-dimensional array_like) or the number of samples available.
+    The confidence intervals are valid if and only if the samples are i.i.d.
+
+    Both one-sided and two-sided confidence intervals can be obtained
+    (default is one-sided). The function returns two values: either the bounds
+    for the two one-sided confidence intervals, or the lower and upper bounds
+    of a two-sided confidence interval. The return values are either the
+    indexes of the bounds (if `x` is an integer) or sample values
+    (if `x` is the set of samples). `None` is returned when there are not
+    enough samples to compute the desired confidence interval.
+
+    There is no uniqueness of the two-sided confidence interval
```

Comment:
I'm confused by "There is no uniqueness ". I'm guessing it means what is is written below: that the two-sided confidence intervals are not necessarily optimal. But when I read it, I interpreted it as meaning that "unique" two-sided confidence intervals can't be found in general.

Reply:
Both are somewhat true. The point here is that, in general, multiple (non-dominated) two-sided CI exists; that is, there may exist several disjoint intervals that contain the quantile of interest with a probability larger than the confidence level. Typically, that would be intervals of the form [x_n, x_m] and [x_n+1, x_m+1]. This happens when the number of samples is large.

Furthermore, there is no obvious "optimal" in this context. Thus, one could return _all_ two-sided CI, but that's a bit cumbersome and, arguably, not really relevant as all CIs tend to be numerically close (i.e., the bound values are similar). So what I propose here is to construct _one_ two-sided interval by composition of two one-sided ones with the same confidence level. This results in a unique CI, but which may be sub-optimal is the sense of: "there may exist a strictly smaller two-sided interval for this confidence level."

I am not sure how far I should go in the level of details in the docstring. What do you think?

Reply:
I think what you have below in the notes is fine. You can probably just remove this here, and mention the bit about one-sided intervals in the notes, too.

Reply:
Sounds good to me.